internal/kafka/internal/workers/kafka_mgrs/kafkas_mgr.go

```diff
@@ -374,3 +387,116 @@ func (k *KafkaManager) updateClusterStatusCapacityMaxMetric(c services.KafkaStre
+			continue
+		}
+
+		instanceSize, err := k.kafkaConfig.GetKafkaInstanceSize(kafka.InstanceType, kafka.SizeId)
+		if err != nil {
+			svcErrors = append(svcErrors, errors.Wrapf(err,
+				"failed to get kafka instance size for %q with instance type %q and size id %q",
+				kafka.ID, kafka.InstanceType, kafka.SizeId,
+			))
+			continue
+		}
+
+		// If lifespan seconds is not defined, expiration is determined based on the user/organisation's active subscription
+		if instanceSize.LifespanSeconds == nil {
+			glog.V(10).Infoln("checking quota entitlement status for Kafka instance %q", kafka.ID)
+			orgBillingModelID := fmt.Sprintf("%s-%s", kafka.OrganisationId, kafka.ActualKafkaBillingModel)
+			active, ok := subscriptionStatusByOrgAndBillingModel[orgBillingModelID]
+			if !ok {
+				isActive, err := k.kafkaService.IsQuotaEntitlementActive(kafka)
+				if err != nil {
+					svcErrors = append(svcErrors, errors.Wrapf(err, "failed to get quota entitlement status of kafka instance %q", kafka.ID))
+					continue
+				}
+				subscriptionStatusByOrgAndBillingModel[orgBillingModelID] = isActive
+				active = isActive
+			}
+
+			if err := k.updateExpiresAtBasedOnQuotaEntitlement(kafka, active); err != nil {
+				svcErrors = append(svcErrors, errors.Wrapf(err, "failed to update expires_at value based on quota entitlement for kafka instance %q", kafka.ID))
+			}
+		} else {
+			// any Kafka instance types with lifespan seconds defined will have their expires_at value set on creation.
+			// this is only temporary to ensure that the expires_at value is set for any Kafka instances created during the rollout of this change.
+			if err := k.updateExpiresAtBasedOnLifespanSeconds(kafka, *instanceSize); err != nil {
+				svcErrors = append(svcErrors, errors.Wrapf(err, "failed to update expires_at value based on lifespanSeconds for kafka instance %q", kafka.ID))
+			}
+		}
+	}
+
+	return svcErrors
+}
+
+// Updates expires_at field of the given Kafka instance based on the user/organisation's quota entitlement status
+func (k *KafkaManager) updateExpiresAtBasedOnQuotaEntitlement(kafka *dbapi.KafkaRequest, isQuotaEntitlementActive bool) error {
+	// if quota entitlement is active, ensure expires_at is set to null
+	if isQuotaEntitlementActive && kafka.ExpiresAt.Valid {
+		glog.V(10).Infof("updating expiration date of kafka instance %q to NULL", kafka.ID)
+		return k.updateKafkaExpirationDate(kafka, nil)
+	}
+
+	// if quota entitlement is not valid and expires_at is not already set, set its value based on the current time and grace period allowance
+	// note that there is a temporary check for zero value of time.Time for cases where the kafka instance expires_at value
+	// hasn't been updated null yet after migration/rollout. This additional check can be removed once all Kafka instances
+	// has been successfully migrated to the new expires_at type.
+	if !isQuotaEntitlementActive && (!kafka.ExpiresAt.Valid || kafka.ExpiresAt.Time.IsZero()) {
```

Comment:
Should we check if `kafka.ExpireAt.Time.After(Now + gracePeriod)` too?

For example, if an instance was set to expire in 10 days, but the subscription has expired today.

I'm not sure that will ever happen though. WDYT?

Reply:
I don't think we need this check 🤔 The only way that the instance expiration is set before the subscription expires is if there was lifespanSeconds defined for that instance type/size. In that case, we always go by the lifespanSeconds lifetime and not change the expiration date even if the grace period doesn't fully complete.